src/hierarchical.rs

```diff
@@ -0,0 +1,493 @@
+use std::cmp::{Ordering, min, max};
+use std::f64::INFINITY;
+use std::mem::replace;
+
+use {Point};
+
+
+/// A function to compute the inter-cluster distance between set `A` and `B`.
+pub type LinkageFunction = Fn(&Fn(usize, usize) -> f64, &[usize], &[usize]) -> f64;
+
+
+/// Hierarchical linkage criteria.
+///
+/// The linkage decides how the distance between clusters is computed.
+#[derive(Copy, Clone)]
+pub enum LinkageCriterion<'a> {
```

Comment:
This problem only exists for custom linkage criteria as I see it and I'm not sure if it's wise to sacrifice expressiveness for a bit of performance in corner cases? It might make more sense to incorporate more linkage criteria directly and to apply known optimizations if the need arises.

What's your opinion on this one?